### PR TITLE
Fix auth via url when password contain unsafe ASCII characters 

### PR DESCRIPTION
**Pull request by sifmelcara**
Before this fix, I cannot auth via url to access http://httpbin.org/basic-auth/user/p@ss
both `simpleHttp "http://user:p@ss@httpbin.org/basic-auth/user/p@ss"`
`simpleHttp http://user:p%40ss@httpbin.org/basic-auth/user/p@ss` will failed
(the former failed due to an exception (InvalidUrlException), the later failed authentication due to encoding p%40ss again).

After this fix, `simpleHttp "http://user:p%40ss@httpbin.org/basic-auth/user/p@ss"` works.

### CODE REVIEW

File: http-client/Network/HTTP/Client/Request.hs [hunk 1]
```diff
@@ -127,8 +127,8 @@ applyAnyUriBasedAuth uri req =
   where
     hasAuth = (notEmpty theuser) && (notEmpty thepass)
     notEmpty = not . null
-    theuser = username authInfo
-    thepass = password authInfo
+    theuser = unEscapeString . takeWhile (/=':') . authPrefix $ authInfo
+    thepass = unEscapeString . takeWhile (/='@') . drop 1 . dropWhile (/=':') $ authInfo
```
Comment (snoyberg) [inline http-client/Network/HTTP/Client/Request.hs]: Why did you change these instead of modifying the implementation of `username` and `password`?

Comment (sifmelcara) [inline http-client/Network/HTTP/Client/Request.hs]: Because I had not notice that `username` and `password` exported here are also used to apply basic auth.
I have update my commits.

**Comment by sifmelcara** [comment]
I updated my commit and refactor the code a little bit.
(By adding a new function `extractBasicAuthInfo`)
`
 ghci> extractBasicAuthInfo $ fromJust $ parseURI "http://user:p%40ss@httpbin.org/user/p@ss"
Just ("user","p@ss")
`

By the way, this code should still work when user did not encode their username/password in URL and their username/password did not cause a parse error in parseUrl and did not contain '%'.

`
 ghci> extractBasicAuthInfo $ fromJust $ parseURI "http://u;ser:pa$$@httpbin.org/u;ser/pa$$"
Just ("u;ser","pa$$")
`

File: http-client/Network/HTTP/Client/Request.hs [hunk 2]
```diff
@@ -121,27 +120,21 @@ getUri req = URI
...
+    case extractBasicAuthInfo uri of
+        Just auth -> uncurry applyBasicAuth auth req
+        Nothing -> req
+
+-- | Extract basic access authentication info in URI.
+-- Return Nothing when there is no auth info in URI.
+extractBasicAuthInfo :: URI -> Maybe (S8.ByteString, S8.ByteString)
+extractBasicAuthInfo uri = do
+    userInfo <- uriUserInfo <$> uriAuthority uri
+    when (':' `notElem` userInfo) Nothing
```
Comment (snoyberg) [inline http-client/Network/HTTP/Client/Request.hs]: More idiomatically:

```haskell
guard (':' `elem` userInfo)
```

File: http-client/Network/HTTP/Client/Request.hs [hunk 3]
```diff
@@ -121,27 +120,21 @@ getUri req = URI
...
+    let username = takeWhile (/=':') userInfo
```
Comment (snoyberg) [inline http-client/Network/HTTP/Client/Request.hs]: It would be more efficient to use `let (username', password') = break (== ':') userInfo`

File: http-client/Network/HTTP/Client/Request.hs [hunk 4]
```diff
@@ -23,8 +23,7 @@ module Network.HTTP.Client.Request
     , setQueryString
     , streamFile
     , observedStreamFile
-    , username
```
Comment (snoyberg) [inline http-client/Network/HTTP/Client/Request.hs]: This broke the test suite, please see the Travis build failure

**Comment by sifmelcara** [comment]
Thank you. I updated my codes and test suite.

**Comment by snoyberg** [comment]
Thanks!
